### DISCUSSION TITLE
CodeReady Containers dev preview for Apple M1 (podman preset only)

### DISCUSSION THREAD

**Discussion by gbraad** [crc-org/crc#2988] [Show and tell]
Here is a mac M1 package of CRC which only works with the Podman preset. If you try it we expect you to provide feedback on tray/UX and/or usability side.

### Instructions
Since it is unsigned binary so you have to put an exception for it. If you somehow installed `aarch64` package previously then make sure you run `sudo rm -fr /Applications/CodeReady\ Containers*`.

During the onboarding please select `podman` because for `openshift` it will not work.

### Download link

  * ~https://storage.googleapis.com/crc-bundle-github-ci/crc-macos-arm64.pkg~  
    Link expired

### Known issues

  * https://github.com/code-ready/tray-electron/issues/53 
  * the terminal doesn't start. Instead use `crc podman-env` from a terminal for now.

**Comment by deboer-tim** [comment]
Here are some issues I found while trying this out, grouped by area:

**Installation/General**
- The initial installer was nice and slick, and followed standard conventions - but just about everything after it needs some polish, text cleaned up, etc.
- The 'branding' was very confusing, especially use of CRC as the term for the traditional CRC, the tray, Podman, the VM, etc. Felt like CRC is the name for anything and everything, which made it harder to tell what it was referring to at any given point.
- Installation progress via output in a terminal window is poor. Terminal should be hidden, maybe parse output to show a normal progress bar.
- Once Podman is running I expect to be able to interact with it in my regular Mac terminal (or other applications).
- With M1 being completely silent you can't tell something is CPU or disk intensive based on fan or disk noise. As a result I tend to have a default assumption that if something isn't instant it must be slow/using tons of memory/CPU/etc. A few times while testing I thought something was bad, but when I did it all again later with CPU/memory monitor open it turned out that everything was surprisingly lightweight. So a) that's awesome! but b) I think we need to bump up the user feedback/animation/etc in a few places.

**System tray**
- No dark mode. :(
- The tray icon is bright red when everything else on the Mac tray is light grey, it doesn't fit in.
- Clicking on the tray icon was glitchy. Sometimes it opens both the popup menu and window, sometimes just one of them, other times it does nothing.
- Looking at other trays, they avoid the problem above by only having one or the other (i.e. a menu with status where you can open a window, or a window that has a menu button if you need more actions). See: Mac tray icons, Viscosity, Sentinel One, RH VPN, etc.
- The tray icon should show the status, e.g. starting, started, etc.

**Notifications**
- No bell/dinging, please. :)
- I don't need notification that I started the tray or that configuration is loaded, these should be removed.
- When I click a button, I don't need notification of it, i.e. I click Start, and it says "hey, someone clicked start!".
- There should be notification when the action is complete (Started), not when it starts (Starting). But with the last point, if I click Start and have the window open watching the status, it's annoying to have the popup notification block my view.
- The one time that notification would be useful is if I click Start and then close the window to do something else while I wait. I'd like to get notified that it's ready, but there's nothing.
- Summary: notification should be reserved for things that happened in the background.

**Configuration**
- The defaults for Podman shouldn't be as big as CRC.
- Things can be set to arbitrary values (e.g. 50 CPU), there doesn't seem to be any validation and changing these doesn't seem to have any effect.

**Popup menu**
- Open developer terminal didn't do anything.
- If I click on the status I get a CRC terminal window. This looks like internal debugging that was left in.

**Window**
- The left part of the window reads as "Status Podman Disk", right side as "Stopped 3.4.4 0%". Needs some formatting.
- Using Docker for years I only saw the version number when upgrading and couldn't tell you what the disk usage was, I don't think these should be on the main screen.
- After it's started I can click Start again (and get another notification). Buttons should be disabled when they can't be used.
- As a new user I click Start, it starts, then what? It wouldn't be obvious what to do next.
- Delete is somewhat ambiguous. As a user the VM is an implementation detail, so the expectation is that it'll completely delete Podman I guess?
- A user would start/stop daily, but rarely Delete, feels like it shouldn't be so prominent.
- Delete should have a confirmation dialog that tells you what it's going to do and the implications (e.g. lose container images).

**Console**
- When it opens it flashes between two screens before it shows the console. This is distracting and makes it feel like it's loading a remote web page instead of a local console.
- Default font size is large. Not picking up system defaults?
- Started downloading a container and happened to click on the Logs tab. It incorrectly says "no container with name or ID "sharp_bhaskara" found: no such container" and "State: Exited".
- If I click on another window and something goes over top there doesn't seem to be any way to get back to the Console without closing the other windows, e.g. it's not on my toolbar or listed as an open window anywhere.
- Don't know what I did, but I clicked in the console trying to search for a container and it went blank. No way to refresh, had to close.
- I clicked on the maximize button accidentally, there is no way to minimize.
- No progress downloading containers, and never seems to finish.
- If I reopen the console, no more containers.

**Reply by gbraad** [reply]
> it feel like it's loading a remote web page instead of a local console.

It opens a remote console inside the VM, based on Cockpit. The flickering hasn't happened visually on Windows, but perhaps it does on macos. It does an authentication flow.

I will pass on some of these comments too, as this is the Container view that is also in RHEL's Console. /cc: @martinpitt. The lack of progress indication is one also that troubled me. Would like to catch up on some of this soon.

**Reply by gbraad** [reply]
> Window
> A user would start/stop daily, but rarely Delete, feels like it shouldn't be so prominent.

Part of a redesign; icon will indicate state of the VM. Delete will show a modal dialog to request deletion with a note about the destructive nature.

  * https://github.com/code-ready/tray-electron/issues/126
  * https://github.com/code-ready/tray-electron/issues/130

**Comment by deboer-tim** [comment]
Ran into a few more issues with the Console:
- If you paste an image name into the Create Container dialog it shows 'No images found' even though the image exists and you can search for and select the same text.
- Depending on how you exit the field, it sometimes deletes what you typed. Or it doesn't delete it (so you now have a form with everything correctly filled out), but Create isn't enabled.
- I got into a state where the Containers section said 'Loading' indefinitely. I tried closing and reopened the Console, stopped and restarted Podman, same thing. I had to delete Podman to clear it.
- Stopping or Deleting Podman doesn't close or update the Console... except sometimes many seconds later. In my case I deleted/recreated/restarted Podman and _then_ it went invalid when I tried to use it again.

**Comment by RiRa12621** [comment]
The app doesn't seem to be signed, hence instant issues on install. I an work around but probably worth noting

![Screenshot 2022-02-07 at 16 32 27](https://user-images.githubusercontent.com/14996407/152819157-b6fb36c8-aef0-493a-b769-77ee4c312985.png)

**Reply by gbraad** [reply]
Dev previews won't be signed, only GA releases. This is also mentioned in the instructions.

**Comment by ecda909** [comment]
My apologies if this question is pretty much implied by the directions and I've missed it somehow... What are the appropriate instructions for running with Podman? I've downloaded the package and installed everything. crc CLI is operating as expected, however when I run crc start I get the hyperkit issue `file '/Users/edermazariegos/.crc/cache/crc_hyperkit_4.9.12.crcbundle' does not exist`... I also get `Unknown preset: openshift` when running `crc setup`

--> I'm quite new to openShift, but I wanted to see if I could chip in and maybe add a different perspective being new to this :)

**Reply by ecda909** [reply]
I've actually just found the GUI. I'll begin the podman setup there instead of the CLI.

**Reply by gbraad** [reply]
Podman preset only
From the CLI you can use:
```
crc podman-env
```

**Reply by ecda909** [reply]
Right, but if I try to start CR with crc from the CLI with the following commands...

`crc setup`
`crc start`

I get 

`Unknown preset: openshift`

It wasn't until I realized that there was a GUI with which I could install CodeReady from. After I completed the installation walkthrough via the GUI, I ran crc podman-env and was able to export the env variables and then the `crc status` command returned metadata on the VM successfully...

`CRC VM:          Running
OpenShift:
Podman:          3.4.4
Disk Usage:      1.847GB of 32.74GB (Inside the CRC VM)
Cache Usage:     34.2GB
Cache Directory: <dir_here>`

**Reply by praveenkumar** [reply]
@ecda909 We only added the podman bundle so `Unknown preset: openshift` might be coming when it try to download the bundle which is not even available. To work with cli you need to do something like following. We encourage user to use GUI with guided workflow.

```
crc config set preset podman
crc setup
crc start
```

**Reply by gbraad** [reply]
This is a dev preview for a reason. It is a very rough release. This is why in the instructions we mentioned:

> During the onboarding please select `podman`

We hope eventually at some point we can offer OpenShift, though for the time being we might wanna be able to remove this for M1. However, this hasn't been implemented as we wanted to create an internal release... which I published publicly nonetheless ;-)

**Reply by gbraad** [reply]
@praveenkumar somehow wondering how this happens:

> We encourage user to use GUI with guided workflow.

the GUI only appears on the restart? Is there a way to just start it after the `pkg` got installed?

**Reply by ecda909** [reply]
Okay gotcha, thanks guys! :)

**Reply by praveenkumar** [reply]
> @praveenkumar somehow wondering how this happens:
> 
> the GUI only appears on the restart? Is there a way to just start it after the `pkg` got installed?

@gbraad No, as soon as you install the pkg you can search Code Ready Containers from finder and it is available.

**Reply by ecda909** [reply]
Yeah, if we can make the GUI appear after installation/restart or tell users what to look for somehow, I think that would be helpful. I didn't know a GUI was available for me to use until I randomly saw it in my Applications folder afterwards.

**Reply by gbraad** [reply]
in Windows you often have a checkbox at the end of an installer that asks `[x] start application after install` and thus when you exit it will start. Not sure if that would be easily possible on macOS. I would suspect the `postinstall` could ... Looking into this. Not that we disallow terminal use, but you shouldn't have to... else the 'onboarding' is failing :-P. Thanks. I hadn't realized this as my macBook is 'kaput' and can't expect a replacement any time soon these days.

**Reply by gbraad** [reply]
* https://github.com/code-ready/crc/issues/2993

**Reply by burrsutter** [reply]
This is the GUI I have seen - https://www.screencast.com/t/t4xSBKqqMF

Is that the right one? Somewhere hidden in there is "podman"?

**Reply by ecda909** [reply]
@burrsutter You have to go through that installation process and then check your Application bin/folder to see the CodeReady Container GUI. It's there that you then specify podman.

**Reply by burrsutter** [reply]
I found "crc-tray", not sure if that is the correct binary but I will give it a run

**Reply by burrsutter** [reply]
I think I found the right GUI
https://www.screencast.com/t/t9Vz0orw

It does not seem appear in the "system tray" using this technique of execution from the terminal

https://www.screencast.com/t/DNLaIf7RHQh

**Reply by burrsutter** [reply]
and the text says
"Your system is correctly setup for using CodeReady Containers, you can now run 'crc start' to start the OpenShift cluster
Starting daemon process ..." when I choose Podman, not OpenShift

**Reply by burrsutter** [reply]
then tried "crc setup" and "crc start"

~/projects/crc-projects $ crc setup
INFO Using bundle path /Users/burr/.crc/cache/crc_podman_hyperkit_3.4.4.crcbundle
INFO Checking if running as non-root
INFO Checking if crc-admin-helper executable is cached
INFO Checking for obsolete admin-helper executable
INFO Checking if running on a supported CPU architecture
INFO Checking minimum RAM requirements
INFO Checking if vfkit is installed
INFO Checking if CodeReady Containers daemon is running
INFO Checking if launchd configuration for tray exists
INFO Check if CodeReady Containers tray is running
INFO Checking if CRC bundle is extracted in '$HOME/.crc'
INFO Checking if /Users/burr/.crc/cache/crc_podman_hyperkit_3.4.4.crcbundle exists
Your system is correctly setup for using CodeReady Containers, you can now run 'crc start' to start the OpenShift cluster
~/projects/crc-projects $ crc start
Is '/Applications/CodeReady Containers.app' running? Cannot reach daemon API: Get "http://unix/api/version": dial unix /Users/burr/.crc/crc-http.sock: connect: connection refused

**Comment by burrsutter** [comment]
I thought this phrase was odd "You can now open CodeReady Containers from application." on the Summary pane of the installer

**Reply by gbraad** [reply]
Text are just placeholders, but this is the line @praveenkumar added to prevent the issue mentioned in https://github.com/code-ready/crc/discussions/2988#discussioncomment-2134367 and https://github.com/code-ready/crc/discussions/2988#discussioncomment-2127300

**Reply by gbraad** [reply]
Added issue

  * #2996

**Comment by burrsutter** [comment]
I do not know what I am doing so I tried:
crc setup
crc start
and that results in 
file '/Users/burr/.crc/cache/crc_hyperkit_4.9.12.crcbundle' does not exist

**Reply by cfergeau** [reply]
`crc config set preset podman` is needed if you don't go through this first time graphical wizard. We'll have to make this the default on m1 builds.

**Reply by burrsutter** [reply]
what is the "graphical wizard"?  I just ran the installer did not see a prompt for "podman"
but I am just rapidly clicking on things :-)

**Reply by gbraad** [reply]
This is the same issue as: https://github.com/code-ready/crc/discussions/2988#discussioncomment-2127300

> We'll have to make this the default on m1 builds.

Default on all macOS builds

**Comment by burrsutter** [comment]
"podman" does seem to be auto-added to my PATH
and I would need the docs that say how to make my normal docker CLI point to the podman daemon vs the docker desktop daemon

**Reply by cfergeau** [reply]
docker compatibility has been worked on in https://github.com/code-ready/crc/pull/3002 but is not part of a release yet. Once this is in, using `eval $(crc podman-env)` will set the `DOCKER_HOST` env var so that `docker` talks to the podman runtime in the VM.

**Reply by gbraad** [reply]
@burrsutter

ideally no additional setup would be needed. No documentation has been added around this yet as this is "dev preview". Also, I do not believe anyone of the documentation team has been assigned/available to work on this as of yet :-/

**Comment by burrsutter** [comment]
Thank you for the update

Just a few notes in a GitHub readme.md describing the happy path steps
would be fine for this alpha/dev preview phase for random testers.

**Comment by burrsutter** [comment]
There are folks like Max Anderson who think docker compatible should be on
out-of-the-box but I do see that as a potential problem for folks with
docker desktop already turned on.  Perhaps the tray GUI could describe to
the user how to address the issue.

**Reply by gbraad** [reply]
Right. This however will be a podman issue first. Let's see how much of an issue this actually is.

For CRC we have chosen to use a different (non conflicting) naming for now. Over time this might change as there are pros/cons for either approach.

Tooling could always be made to detect and handle this based on user's input.

Note: not only Max has this opinion. Several people have suggested this as part of the investigation around desktop tooling. An easy onboarding and OOTB experience seems a necessity. I think 'detecting conflicts' might be a possible solution to still allow this.

**Reply by veillard** [reply]
agree with Gerard

**Comment by pvalena** [comment]
The link gives me access denied.

**Reply by gbraad** [reply]
You mean the download? If so, yes. This expired as this was merely an early preview.

**Reply by pvalena** [reply]
Yes. So... is it developed or tracked somewhere? I came here from some PR, but I lost it now:) ... I might be able to give some feedback / debug it, when I get to it. (I don't need it now, but I think I'll need to use it the future.)

_ _ _ _ 

_Thanks for the reply! and sorry for my previous brevity, it was a bit late for me..._

**Comment by ecda909** [comment]
Is the dev preview link no longer working?

**Reply by pvalena** [reply]
```
Request URL: https://storage.googleapis.com/crc-bundle-github-ci/crc-macos-arm64.pkg
Request Method: GET
Status Code: 403 
```
_""This expired as this was merely an early preview.""_

**Reply by ecda909** [reply]
Ah, is there another package we can work off of?

**Reply by gbraad** [reply]
We are in the process to move to vfkit as tbe default driver on maxOS. Until this is finalized no new version will be published. We hope to conclude this in the nextone or two sprints.

**Reply by gbraad** [reply]
The vfkit has been made default in the current release. No negative feedback came back as of yet.
